File: src/api/subsonic/types/song.ts
import * as v from 'valibot';

const ReplayGain = v.partial(
  v.object({
    albumGain: v.number(),
    albumPeak: v.number(),
    trackGain: v.number(),
    trackPeak: v.number(),
  }),
);

export const Song = v.object({
  album: v.string(),
  albumId: v.string(),
  artist: v.string(),
  artistId: v.optional(v.string()),
  coverArt: v.string(),
  discNumber: v.optional(v.number()),
  duration: v.number(),
  genre: v.optional(v.string()),
  genres: v.optional(v.array(v.object({ name: v.string() }))),
  id: v.string(),
  musicBrainzId: v.optional(v.string()),
  replayGain: v.optional(ReplayGain),
  size: v.number(),
  starred: v.optional(v.string()),
  title: v.string(),
  track: v.optional(v.number()),
  userRating: v.optional(v.number()),
  year: v.optional(v.number()),
});

export type Song = v.InferInput<typeof Song>;

File: src/api/subsonic/song.ml
open Ppx_deriving_json_runtime.Primitives

type[@deriving json] t = {
  album : string;
  albumId : string;
  artist : string;
  artistId : string option; [@json.option]
  coverArt : string;
  discNumber : int option; [@json.option]
  duration : float;
  genre : string option; [@json.option]
  genres : Genre.t array option; [@json.option]
  id : string;
  musicBrainzId : string option; [@json.option]
  replayGain : ReplayGain.t option; [@json.option]
  size : float;
  starred : string option; [@json.option]
  title : string;
  track : int option; [@json.option]
  userRating : float option; [@json.option]
  year : int option; [@json.option]
}
